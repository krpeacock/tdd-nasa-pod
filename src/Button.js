import React, {Component} from 'react'
export function Button (props) {
  return (
    <button onClick={props.onClickCallback}>{props.children}</button>
  )
}