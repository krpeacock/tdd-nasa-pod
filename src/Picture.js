import React, {Component} from 'react'
import { fetchNasaPicture } from "./fetchNasaPicture";

export class Picture extends Component {
  state = {
    title: "Title",
    url: "https://picsum.photos/200/300",
    explanation: "Description"
  }

  async componentWillMount(){
    const data = await this.props.fetchNasaPicture(this.props.date)
    this.setState({...data})
  }

  async componentWillReceiveProps(nextProps) {
    console.log(nextProps)
    const data = await nextProps.fetchNasaPicture(nextProps.date)
    this.setState(data)
  }

  render(){
    let {title, url, explanation} = this.state
    return (
      <div>
        <h1 className="title">{title}</h1>
        <img src={url} />
        <p className="explanation">{explanation}</p>
      </div>
    )
  }
}

Picture.defaultProps = {
  fetchNasaPicture: fetchNasaPicture
}