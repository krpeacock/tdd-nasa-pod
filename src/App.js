import React, { Component } from 'react';
import {Button} from './Button';
import {Picture} from "./Picture";
import {dateToText} from './dateToText';
import logo from './logo.svg';
import './App.css';

const millisecondsPerDay = 24 * 60 * 60 * 1000
const directions = {
  BACK: -millisecondsPerDay,
  FORWARD: millisecondsPerDay
}

class App extends Component {
  state = {
    date: new Date()
  }

  onDecreaseDate = () => {
    this.stepByDay('BACK')
  }

  onIncreaseDate = () => {
    this.stepByDay('FORWARD')
  }

  stepByDay = (direction) => {
    let newDate = new Date(this.state.date.getTime() + directions[direction])
    if (newDate <= Date.now()){
      this.setState({date: newDate})
    }
  }

  render() {
    return (
      <div className="App">
        <h1 className="header">Nasa Picture of the Day</h1>
        <Button className="test-back" onClickCallback={this.onDecreaseDate}>{'\<'}</Button>
        <span className="date">{dateToText(this.state.date)}</span>
        <Button className="test-forward" onClickCallback={this.onIncreaseDate}>{'\>'}</Button>
        <Picture date={dateToText(this.state.date)}/>
      </div>
    );
  }
}

export default App;
