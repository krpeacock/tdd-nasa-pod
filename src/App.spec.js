import React from 'react';
import Enzyme, {shallow, mount} from 'enzyme';
import App from './App'
import {Button} from './Button';
import { Picture } from "./Picture";
import adapter from 'enzyme-adapter-react-16';
import MockDate from 'mockdate';
import fetchMock from 'fetch-mock'
import {createSerializer} from 'enzyme-to-json';

Enzyme.configure({adapter: new adapter() });
expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));

describe('App', () => {
  beforeEach(()=>{
    MockDate.reset()
  })

  describe('render', ()=> {
    it('should render the header', () => {
      const wrapper = shallow(<App/>)

      const header = wrapper.find('h1.header')
      expect(header.length).toEqual(1)
      expect(header.text()).toEqual("Nasa Picture of the Day")
    })
    it('should render the current date in a human-readable format', () => {
      MockDate.set('1/2/2018');
      const wrapper = shallow(<App/>);
      const dateSpan = wrapper.find('span.date');
      expect(dateSpan.length).toEqual(1);
      expect(dateSpan.text()).toEqual('2018-01-02');
    })

    it('should render two buttons', ()=>{
      const wrapper = shallow(<App/>);
      const button = wrapper.find(Button)
      expect(button.length).toEqual(2)
    })

    it('should render a button for moving back in time', ()=>{
      const wrapper = shallow(<App/>);
      const button = wrapper.find(Button).filterWhere(b=>b.children().text()==="<")
      expect(button.length).toEqual(1)
    })

    it('should render a button for moving forward in time', ()=>{
      const wrapper = shallow(<App/>);
      const button = wrapper.find(Button).filterWhere(b=>b.children().text()===">")
      expect(button.length).toEqual(1)
    })

    it('should render the picture component with the date from the state passed to it', () => {
      MockDate.set("4/5/2010")
      const wrapper = shallow(<App/>)
      const picture = wrapper.find(Picture).find({date: "2010-04-05"})
      expect(picture.length).toEqual(1);
    })

    it('renders correctly', ()=> {
      const wrapper = shallow(<App/>)
      MockDate.set("4/5/2010")
      expect(wrapper).toMatchSnapshot()
    })

    describe('change date',()=>{
      it('should change date back when clicked on the "back in time" button', ()=>{
        MockDate.set('2/3/2017')
        const wrapper = mount(<App/>)
        const dateSpan = wrapper.find('span.date')
        const backInTimeButton = wrapper.find(Button).filter('.test-back')

        expect(dateSpan.text()).toEqual("2017-02-03")
        backInTimeButton.simulate('click')

        expect(dateSpan.text()).toEqual("2017-02-02")
      })
      it('should change date forward when you click on the "forward in time" button', ()=>{
        MockDate.set('2/3/2017')
        const wrapper = mount(<App/>)
        const dateSpan = wrapper.find('span.date')
        const backInTimeButton = wrapper.find(Button).filter('.test-back')

        expect(dateSpan.text()).toEqual("2017-02-03")
        
        backInTimeButton.simulate('click')

        expect(dateSpan.text()).toEqual("2017-02-02")

        const forwardInTimeButton = wrapper.find(Button).filter('.test-forward')

        forwardInTimeButton.simulate('click')

        expect(dateSpan.text()).toEqual("2017-02-03")
      })

    })
  })
})