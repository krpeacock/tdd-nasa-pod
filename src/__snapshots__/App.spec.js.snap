// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render renders correctly 1`] = `
<div
  className="App"
>
  <h1
    className="header"
  >
    Nasa Picture of the Day
  </h1>
  <Button
    className="test-back"
    onClickCallback={[Function]}
  >
    &lt;
  </Button>
  <span
    className="date"
  >
    2018-02-06
  </span>
  <Button
    className="test-forward"
    onClickCallback={[Function]}
  >
    &gt;
  </Button>
  <Picture
    date="2018-02-06"
    fetchNasaPicture={[Function]}
  />
</div>
`;
